### PR TITLE
HIVE-27117: Fix compaction related flaky tests

### PR DESCRIPTION
**Pull request by veghlaci05**
### What changes were proposed in this pull request?
Fix the flaky compaction tests
Note: follow-up task: https://issues.apache.org/jira/browse/HIVE-27121

### Why are the changes needed?
In order to make CI builds more stable

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manually

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorOnTezTest.java [hunk 1]
```diff
@@ -185,23 +187,32 @@ protected void verifySuccessfulCompaction(int expectedSuccessfulCompactions) thr
...
-    for (ProtoMessageReader<HiveHookEvents.HiveHookEventProto> reader : readers) {
-      HiveHookEvents.HiveHookEventProto event = reader.readEvent();
-      boolean getRelatedEvent = false;
-      while (!getRelatedEvent) {
-        while (event != null && ExecutionMode.TEZ != ExecutionMode.valueOf(event.getExecutionMode())) {
-          event = reader.readEvent();
+    int retryCount = 3;
+    while (retryCount-- > 0) {
+      try {
+        List<ProtoMessageReader<HiveHookEvents.HiveHookEventProto>> readers = TestHiveProtoLoggingHook.getTestReader(conf, tmpFolder);
```
Comment (abstractdog) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorOnTezTest.java:193]: is it possible to write a countdownlatch magic here similarly to the fix in TestTxnCommands2

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java [hunk 2]
```diff
@@ -3647,9 +3647,12 @@ long generateCompactionQueueId(Statement stmt) throws SQLException, MetaExceptio
             + "no record found in next_compaction_queue_id");
       }
       long id = rs.getLong(1);
-      s = "UPDATE \"NEXT_COMPACTION_QUEUE_ID\" SET \"NCQ_NEXT\" = " + (id + 1);
+      s = "UPDATE \"NEXT_COMPACTION_QUEUE_ID\" SET \"NCQ_NEXT\" = " + (id + 1) + " WHERE \"NCQ_NEXT\" = " + id;
       LOG.debug("Going to execute update <{}>", s);
-      stmt.executeUpdate(s);
+      if (stmt.executeUpdate(s) != 1) {
```
Comment (abstractdog) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3652]: this looks like a hack in production code while we're targeting test fixes here, which is a bit concerning
this method is already very confusing: it looks like cannot guarantee synchronization regarding NCQ_NEXT (which is okay if we have more instances running this code, right?), as we're selecting it, doing something here in java code (increment), then trying to write it back, isn't there a way to use AUTOINCREMENTed value across the supported RDBMS types?
I mean, this recursive call is not something I really like, we should find a better synchronization pattern, I think a compaction queue id increment is not something that happens 100 times per second, so even a RDBMS-level lock is fine on this table I guess (I don't know this area though)

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorOnTezTest.java (hunk 1 above)
Comment (veghlaci05) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/CompactorOnTezTest.java:193]: Unofrtunately not possible, the problem here that I cannot mock the Event processing part to  release the latch.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java (hunk 2 above)
Comment (veghlaci05) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3652]: This is a short term quick and dirty fix, as the id generation for compaction requests is broken. The old approach did not guarantee a unique id at all, and since Initiator processes the candidates in a parallel way, this could be a production issue too. Actually I have seen this duplicate key error in production logs before. The long term solution is to modify the id to auto increment field. It is already done this way for TXNS table. However, this change will be bigger, therefore I created a separate task for it: https://issues.apache.org/jira/browse/HIVE-27121

Comment (abstractdog) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java:3652]: okay, we can live with a follow-up task, but in this case, I would like to see further minor improvements here:
- HIVE-27121 <- refer to the follow-up task for better reference in a code comment
- use LOG.info at least, this is an unhappy codepath, debug level is not enough
- show the current id in the log message for better reference
-

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java [hunk 3]
```diff
@@ -577,7 +577,7 @@ public RawStore getMS() throws MetaException {
     return getMSForConf(conf);
   }
 
-  public static RawStore getMSForConf(Configuration conf) throws MetaException {
+  public static synchronized RawStore getMSForConf(Configuration conf) throws MetaException {
```
Comment (abstractdog) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:580]: can you please clarify how is this synchronization related to the solution?

Comment (veghlaci05) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HMSHandler.java:580]: This is another flakyness. Without this, org.apache.hadoop.hive.ql.txn.compactor.TestCompactionMetrics.testInitiatorFailuresCountedCorrectly deadlocked repeatedly on my machine. 4 threads were trying to init the DB (and get the RawStore) concurrently at the smae time which resulted in a deadlock.

File: ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands2.java [hunk 4]
```diff
@@ -1434,6 +1434,8 @@ public void testTruncatePartitionAndCompactionConcurrent() throws Exception {
     execDDLOpAndCompactionConcurrently("TRUNCATE_PARTITION", true);
   }
   private void execDDLOpAndCompactionConcurrently(String opType, boolean isPartioned) throws Exception {
+    hiveConf.setBoolVar(HiveConf.ConfVars.HIVE_COMPACTOR_GATHER_STATS, false);
```
Comment (abstractdog) [inline ql/src/test/org/apache/hadoop/hive/ql/TestTxnCommands2.java:1437]: please comment on why is this needed and then it can be merged +1

**Review by abstractdog — APPROVED** [review]
